pkg/threescale/backend/manager.go

```diff
@@ -0,0 +1,370 @@
+type Manager struct {
+	// Connections is a map of connections to the backend where
+	// string is the URL of the threescale backend
+	Connections map[string]Backend
+}
+
+// Backend defines the connection to a single backend and maintains a cache
+// for multiple services and applications per backend. It implements the 3scale Client interface
+type Backend struct {
+	client threescale.Client
+	cache  Cacheable
+}
+
+// Application defined under a 3scale service
+type Application struct {
+	LastResponse     *threescale.AuthorizeResult
+	LimitCounter     LimitCounter
+	UnlimitedCounter map[string]int
+}
+
+// Limit captures the current state of the rate limit for a particular time period
+type Limit struct {
+	current int
+	max     int
+}
+
+// LimitCounter keeps a count of limits for a given period
+type LimitCounter map[string]map[api.Period]*Limit
+
+// Authorize authorizes a request based on the current cached values
+// If the request misses the cache, a remote call to 3scale is made
+// Request Transactions must not be nil and must not be empty
+// If multiple transactions are provided, all but the first is discarded
+func (b *Backend) Authorize(request threescale.Request) (*threescale.AuthorizeResult, error) {
+	err := b.validateTransaction(request.Transactions)
+	if err != nil {
+		return nil, err
+	}
+
+	app, _, err := b.getApplicationFromRequest(request, request.Transactions[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to process request - %s", err.Error())
+	}
+
+	affectedMetrics := b.computeAffectedMetrics(app.LastResponse.Hierarchy, request.Transactions[0].Metrics)
+	_, isAuthorized := b.isAuthorized(app, affectedMetrics)
+
+	result := &threescale.AuthorizeResult{Authorized: isAuthorized}
+	if !isAuthorized {
+		result.ErrorCode = "usage limits are exceeded"
+	}
+
+	return result, nil
+
+}
+
+// AuthRep authorizes a request based on the current cached values and reports
+// the new state to the cache in cases where the request has been authorized
+// If the request misses the cache, a remote call to 3scale is made
+// Request Transactions must not be nil and must not be empty
+// If multiple transactions are provided, all but the first is discarded
+func (b *Backend) AuthRep(request threescale.Request) (*threescale.AuthorizeResult, error) {
+	err := b.validateTransaction(request.Transactions)
+	if err != nil {
+		return nil, err
+	}
+
+	app, cacheKey, err := b.getApplicationFromRequest(request, request.Transactions[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to process request - %s", err.Error())
+	}
+
+	affectedMetrics := b.computeAffectedMetrics(app.LastResponse.Hierarchy, request.Transactions[0].Metrics)
+	clonedApplication, isAuthorized := b.isAuthorized(app, affectedMetrics)
+
+	result := &threescale.AuthorizeResult{Authorized: isAuthorized}
+	if isAuthorized {
+		b.cache.Set(cacheKey, clonedApplication)
+	} else {
+		result.ErrorCode = "usage limits are exceeded"
+	}
```

Comment:
We will also want to cache the limits exceeded case via some flag so we don't need to recompute stuff, but that's secondary at this point (also applies to `Authorize`).

Reply:
gotcha, will add a boolean value to the the structure and mark it as exceeded to prevent iterating further in that cycle